modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/QueryCancel.java

```diff
@@ -17,45 +17,134 @@
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
 import org.apache.ignite.internal.lang.IgniteInternalException;
 import org.apache.ignite.internal.util.Cancellable;
+import org.jetbrains.annotations.Nullable;
 
 /**
  * Holds query cancel state.
  */
 public class QueryCancel {
     private final List<Cancellable> cancelActions = new ArrayList<>(3);
 
-    private boolean canceled;
+    private Reason reason;
+
+    private volatile Instant deadline;
+
+    private volatile CompletableFuture<Void> timeoutFut;
 
     /**
-     * Adds a cancel action.
+     * Adds a cancel action. If operation has already been canceled, throws a {@link QueryCancelledException}.
+     *
+     * <p>NOTE: If the operation is cancelled, this method will immediately invoke the given action
+     * and then throw a {@link QueryCancelledException}.
      *
-     * @param clo Add cancel action.
+     * @param clo Cancel action.
      */
     public synchronized void add(Cancellable clo) throws QueryCancelledException {
         assert clo != null;
 
-        if (canceled) {
-            throw new QueryCancelledException();
+        if (reason != null) {
+            boolean timeout = reason == Reason.TIMEOUT;
+
+            // Immediately invoke a cancel action, if already cancelled.
+            // Otherwise the caller is required to catch QueryCancelledException and call an action manually.
+            try {
+                clo.cancel(timeout);
+            } catch (Exception ignore) {
+                // Do nothing
+            }
+
+            String message = timeout ? QueryCancelledException.TIMEOUT_MSG : QueryCancelledException.CANCEL_MSG;
+            throw new QueryCancelledException(message);
         }
 
         cancelActions.add(clo);
     }
 
+    /**
+     * Removes the given callback.
+     *
+     * @param clo Callback.
+     */
+    public synchronized void remove(Cancellable clo) {
+        assert clo != null;
+
+        cancelActions.remove(clo);
+    }
+
+    /**
+     * Schedules a timeout after {@code timeoutMillis} milliseconds.
+     * Call be called only once.
+     *
+     * @param scheduler Scheduler to trigger an action.
+     * @param timeoutMillis Timeout in milliseconds.
+     * @return Future that will be completed when the timeout is reached.
+     */
+    public synchronized CompletableFuture<Void> setTimeout(ScheduledExecutorService scheduler, long timeoutMillis) {
+        assert reason == null : "Cannot set a timeout when cancelled";
+        assert timeoutFut == null : "Timeout has already been set";
+
+        CompletableFuture<Void> fut = new CompletableFuture<>();
+        fut.thenAccept((r) -> doCancel(Reason.TIMEOUT));
+
+        ScheduledFuture<?> f = scheduler.schedule(() -> {
+            fut.complete(null);
+        }, timeoutMillis, MILLISECONDS);
+
+        add((timeout) -> {
+            // Cancel the future if we didn't timeout.
```

Comment:
```suggestion
            // Cancel the future if we didn't timeout,
            // since in the case of a timeout it is already completed.
```
:thinking:

Reply:
Done.